src/Integration/MefServices/ProjectsRuleSetProvider.cs

```diff
@@ -0,0 +1,355 @@
+        {
+            if (projectSystemHelper == null)
+            {
+                throw new ArgumentNullException(nameof(projectSystemHelper));
+            }
+
+            if (fileChangeService == null)
+            {
+                throw new ArgumentNullException(nameof(fileChangeService));
+            }
+
+            if (activeSolutionTracker == null)
+            {
+                throw new ArgumentNullException(nameof(activeSolutionTracker));
+            }
+
+            if (configurationProvider == null)
+            {
+                throw new ArgumentNullException(nameof(configurationProvider));
+            }
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            this.projectSystemHelper = projectSystemHelper;
+            this.fileChangeService = fileChangeService;
+            this.activeSolutionTracker = activeSolutionTracker;
+            this.configurationProvider = configurationProvider;
+            this.logger = logger;
+
+            this.activeSolutionTracker.ActiveSolutionChanged += OnActiveSolutionChanged;
+            this.activeSolutionTracker.AfterProjectOpened += OnAfterProjectOpened;
+
+            BuildCacheAsync();
+        }
+
+        private async void OnAfterProjectOpened(object sender, ProjectOpenedEventArgs e)
+        {
+            if (this.projectSystemHelper.IsSolutionFullyOpened() &&
+                e.Project != null &&
+                !string.IsNullOrEmpty(e.Project.FullName) &&
+                !projectPathToCachedData.ContainsKey(e.Project.FullName))
+            {
+                // Improve me: there is no need to rebuild the full cache here
+                ClearCache();
+                await BuildCacheAsync();
+            }
+        }
+
+        private async void OnActiveSolutionChanged(object sender, ActiveSolutionChangedEventArgs e)
+        {
+            ClearCache();
+
+            if (e.IsSolutionOpen)
+            {
+                await BuildCacheAsync();
+            }
+        }
+
+        public bool HasRuleSetWithSonarAnalyzerRules(string projectFilePath) =>
+            projectPathToCachedData.ContainsKey(projectFilePath) && projectPathToCachedData[projectFilePath].HasAnySonarRule;
+
+        private void ClearCache()
+        {
+            filesToTrack.Values.ToList().ForEach(x => x.UpdatedOnDisk -= OnUpdatedOnDisk);
+            filesToTrack.Clear();
+
+            projectPathToCachedData.Clear();
+        }
+
+        private async Task BuildCacheAsync()
+        {
+            while (!projectSystemHelper.IsSolutionFullyOpened())
+            {
+                await Task.Delay(2000); // REVIEW: Is 2s a good enough time?
+            }
+
+            projectSystemHelper.GetSolutionProjects()
+                .ToList()
+                .ForEach(p => projectPathToCachedData.Add(p.FullName, ProcessProjectAndCollectFilesToTrack(p)));
```

Comment:
We should really add an extension method that gives us a _ForEach_ over an enumeration so we don't always have to put it in a list first.